src/main/flight/wind_estimator.c

```diff
@@ -0,0 +1,155 @@
+    float groundVelocityDiff[XYZ_AXIS_COUNT];
+    float groundVelocitySum[XYZ_AXIS_COUNT];
+
+    float fuselageDirection[XYZ_AXIS_COUNT];
+    float fuselageDirectionDiff[XYZ_AXIS_COUNT];
+    float fuselageDirectionSum[XYZ_AXIS_COUNT];
+
+    // Get current 3D velocity from GPS in cm/s
+    // relative to earth frame
+    groundVelocity[X] = gpsSol.velNED[X];
+    groundVelocity[Y] = gpsSol.velNED[Y];
+    groundVelocity[Z] = gpsSol.velNED[Z];
+
+    // Fuselage direction in earth frame
+    fuselageDirection[X] = rMat[0][0];
+    fuselageDirection[Y] = rMat[1][0];
+    fuselageDirection[Z] = rMat[2][0];
+
+    timeDelta_t timeDelta = cmpTimeUs(currentTimeUs, lastUpdateUs);
+    // scrap our data and start over if we're taking too long to get a direction change
+    if (lastUpdateUs == 0 || timeDelta > 2 * USECS_PER_SEC) {
+
+        lastUpdateUs = currentTimeUs;
+
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+        return;
+    }
+
+    fuselageDirectionDiff[X] = fuselageDirection[X] - lastFuselageDirection[X];
+    fuselageDirectionDiff[Y] = fuselageDirection[Y] - lastFuselageDirection[Y];
+    fuselageDirectionDiff[Z] = fuselageDirection[Z] - lastFuselageDirection[Z];
+
+    float diff_length = sqrtf(sq(fuselageDirectionDiff[X]) + sq(fuselageDirectionDiff[Y]) + sq(fuselageDirectionDiff[Z]));
+    // Very small changes in attitude will result in a denominator
+    // very close to zero which will introduce too much error in the
+    // estimation.
+    //
+    // TODO: Is 0.2f an adequate threshold?
+    if (diff_length > 0.2f) {
+        // when turning, use the attitude response to estimate wind speed
+        groundVelocityDiff[X] = groundVelocity[X] - lastGroundVelocity[X];
+        groundVelocityDiff[Y] = groundVelocity[Y] - lastGroundVelocity[Y];
+        groundVelocityDiff[Z] = groundVelocity[X] - lastGroundVelocity[Z];
+
+        // estimate airspeed it using equation 6
+        float V = (sqrtf(sq(groundVelocityDiff[0]) + sq(groundVelocityDiff[1]) + sq(groundVelocityDiff[2]))) / diff_length;
+
+        fuselageDirectionSum[X] = fuselageDirection[X] + lastFuselageDirection[X];
+        fuselageDirectionSum[Y] = fuselageDirection[Y] + lastFuselageDirection[Y];
+        fuselageDirectionSum[Z] = fuselageDirection[Z] + lastFuselageDirection[Z];
+
+        groundVelocitySum[X] = groundVelocity[X] + lastGroundVelocity[X];
+        groundVelocitySum[Y] = groundVelocity[Y] + lastGroundVelocity[Y];
+        groundVelocitySum[Z] = groundVelocity[Z] + lastGroundVelocity[Z];
+
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+
+        float theta = atan2f(groundVelocityDiff[1], groundVelocityDiff[0]) - atan2f(groundVelocityDiff[1], groundVelocityDiff[0]);// equation 9
```

Comment:
as far as I understand from documentation here should be the difference between vectors of ground velocity speed and fuselage direction. I think it will work:

`
float theta = atan2f(groundVelocityDiff[1], groundVelocityDiff[0]) - atan2f(fuselageDirectionDiff[1], fuselageDirectionDiff[0]);
`

Reply:
Yes, I didn't see the document, but your suggestion looks more correct than the original code.